### PR TITLE
Fix workflow refresh for closed workflows

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Allow workflow refresh for closed workflows
- Remove duplicated/buggy GetCloseEvent() method in mutableState
- Remove unused parameter from workflow refresher, task generator, mutableState and timerSequence
- Handle get history client error when calling RefreshWorkflowTask 
- Remove defer() in CapturePanic() as this function is already called in defer

**Why?**
- Bug fix for refresh closed workflows

**How did you test it?**
- Tested locally

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

**Review by longquanzheng — APPROVED** [review]

**Review by yux0 — APPROVED** [review]

File: service/history/execution/mutable_state_task_generator.go [hunk 1]
```diff
@@ -145,10 +136,11 @@ func (r *mutableStateTaskGeneratorImpl) GenerateWorkflowStartTasks(
 
 	executionInfo := r.mutableState.GetExecutionInfo()
 	startVersion := startEvent.GetVersion()
+	startTimestamp := time.Unix(0, startEvent.GetTimestamp())
 
 	workflowTimeoutDuration := time.Duration(executionInfo.WorkflowTimeout) * time.Second
 	workflowTimeoutDuration = workflowTimeoutDuration + firstDecisionDelayDuration
-	workflowTimeoutTimestamp := now.Add(workflowTimeoutDuration)
+	workflowTimeoutTimestamp := startTimestamp.Add(workflowTimeoutDuration)
```
Comment (yycptt) [inline service/history/execution/mutable_state_task_generator.go:143]: Looks like we still need to pass in a timestamp here since workflow reset needs to use reset time instead of start event time for calculating the timeout timestamp. 
Will revert the changes for this function to preserve the now parameter.

Comment (longquanzheng) [reply service/history/execution/mutable_state_task_generator.go]: How about change the parameter name to something else. It’s not really now. Also a comment would help understand why we don’t use start event timestamp

Comment (longquanzheng) [reply service/history/execution/mutable_state_task_generator.go]: Eg: timeoutStartTime

Comment (yycptt) [reply service/history/execution/mutable_state_task_generator.go]: sure. I renamed it to `startTime` and added a comments says for reset the workflow start time should be the reset time in order to re-calculated the timeout timestamp.

**Comment by ceearrashee** [comment]
Hello!
The fix that you made did not include in the last releases. 
Can you guide me on when this fix will be released and in which release?

**Comment by longquanzheng** [comment]
@ceearrashee Hello, it's included in the next minor release branch: https://github.com/uber/cadence/commits/0.23.x
v0.23.0 is being baked and will be out soon. Stay tuned.
